_drafts/debug-monitor.md

```diff
@@ -0,0 +1,676 @@
+
+For the NRF52, we can install a breakpoint using the following logic:
+
+```c
+bool fpb_set_breakpoint(size_t comp_id, uint32_t instr_addr) {
+  if (instr_addr >= 0x20000000) {
+    // for revision 1 only breakpoints in code can be installed :/
+    return false;
+  }
+  // make sure the FPB is enabled
+  FPB->FP_CTRL |= 0x3;
+
+  const uint32_t replace = (instr_addr & 0x2) == 0 ? 1 : 2;
+  const uint32_t fp_comp = (instr_addr & ~0x3) | 0x1 | (replace << 30);
+  FPB->FP_COMP[comp_id] = fp_comp;
+  return true;
+}
+```
+
+> For more details on how to configure the Flash Patch And Breakpoint (**FPB**) unit,
+> [check out our post on the topic]({% post_url 2020-06-17-cortex-m-breakpoints %}#flash-patch--breakpoint-unit).
+
+To facilitate breakpoint testing, I've added `dummy_function_1` through `dummy_function_9` to my
+program. All the functions do is print the name of the function being executed:
+
+```c
+void dummy_function_3(void) {
+  EXAMPLE_LOG("stub function '%s' called", __func__);
+}
+```
+
+We will be calling these functions to test out or debug monitor mode implementation.
+
+I've also added a few CLI commands to the shell to help us configure the DebugMonitor exception
+handler, enable breakpoints, and call the dummy functions:
+
+```bash
+shell> help
+bkpt: Issue a Breakpoint Instruction
+debug_mon_en: Enable Monitor Debug Mode
+fpb_dump: Dump Active FPB Settings
+fpb_set_breakpoint: Set Breakpoint [Comp Id] [Address]
+call_dummy_funcs: Invoke dummy functions
+dump_dummy_funcs: Print first instruction of each dummy function
+help: Lists all commands
+```
+
+Let's try configuring a breakpoint for `dummy_function_4`, address `0x270`, and see what happens:
+
+```bash
+shell> dump_dummy_funcs
+ ...
+dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
+ ...
+shell> debug_mon_en
+Monitor Mode Debug Enabled!
+
+shell> fpb_set_breakpoint 0 0x270
+Set breakpoint on address 0x270 in FP_COMP[0] Succeeded
+
+shell> fpb_dump
+FPB Revision: 0, Enabled: 1, Hardware Breakpoints: 6
+  FP_COMP[0] Enabled 1, Replace: 1, Address 0x270
+  ...
+
+shell> call_dummy_funcs
+stub function 'dummy_function_1' called
+stub function 'dummy_function_2' called
+stub function 'dummy_function_3' called
+DebugMonitor Exception
+DEMCR: 0x01010000
+DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
+Register Dump
+ r0  =0x200031d8
+ r1  =0x00000028
+ r2  =0x00000001
+ r3  =0x00000271
+ r12 =0xffffffff
+ lr  =0x000009ed
+ pc  =0x00000270
+ xpsr=0x010f0000
+Debug Event Detected, Awaiting 'c' or 's'
+```
+
+As expected we see our `DebugMon_Handler` has been invoked on address `pc =0x00000270` which
+is where we installed the breakpoint!
+
+#### Adding a CLI to the DebugMonitor Exception
+
+To add a CLI for DebugMonitor handling, all we need is a way to read in bytes over a UART from a ISR. This
+can be done by either polling the peripheral or by using an interrupt that is configured to run at
+a higher priority than the DebugMonitor exception to do the reading.
+
```

Comment:
Mention the shell library?

Reply:
We'd have to refactor the shell a little bit to support this use case. There's only one context so it would not be safe to have the user land shell and the ISR shell share the same thing